node/follower_node.go

```diff
@@ -0,0 +1,390 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+// Package node is the Algorand node itself, with functions exposed to the frontend
+package node
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/algorand/go-algorand/agreement"
+	"github.com/algorand/go-algorand/catchup"
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data"
+	"github.com/algorand/go-algorand/data/account"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/ledger"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/network"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/rpcs"
+	"github.com/algorand/go-algorand/util/execpool"
+)
+
+// AlgorandFollowerNode implements follower mode/ledger delta APIs and disables participation-related methods
+type AlgorandFollowerNode struct {
+	AlgorandFullNode
```

Comment:
Yes, I'm aware of this.

I'll take a look, but the reason I think embedding is better in this case is reducing duplication of implementations in the case where we're not altering things. Any spot where implementation gets redefined presents the possibility of `node.go` getting updated, but not our duplicate implementation.

The proper go way to handle these things is to split everything out into separate interfaces, but I'm trying to avoid that for obvious reasons. Possibly the amount of methods I've had to redefine means the embedding is not actually doing that much for us anymore.

Reply:
No, there's ~7 functions we get to reuse for free. As long as you don't redefine a variable things just work.
https://play.golang.com/p/kmRkK4e9wfj 


In any case the reuse is not so much that it makes a huge difference. I somewhat see the value of completely separating these implementations.